File: src/csharp/LogosSort.cs
using System;
using System.Runtime.CompilerServices;

namespace LogosSortLib
{
    /// <summary>
    /// LogosSort — Golden-ratio dual-pivot introsort (C#)
    ///
    /// Two pivots placed at φ (≈61.8%) and 1−φ (≈38.2%) golden-ratio positions
    /// of a chaos-seeded index. Ninther pivot refinement, three-way partition,
    /// counting-sort fast path for dense int ranges, O(log n) depth limit.
    /// </summary>
    public static class LogosSort
    {
        private const double Phi  = 0.6180339887498949;
        private const double Phi2 = 0.3819660112501051;
        private const int    SmallN = 48;

        [ThreadStatic]
        private static Random? _rng;
        private static Random Rng => _rng ??= new Random();

        // ─────────────────────────────────────────────────────────────────────
        // Public API
        // ─────────────────────────────────────────────────────────────────────

        /// <summary>Sort <paramref name="arr"/> in place.</summary>
        public static void Sort(int[] arr)
        {
            int n = arr.Length;
            if (n < 2) return;
            int depthLimit = 2 * BitLength(n) + 4;
            SortImpl(arr, 0, n - 1, depthLimit);
        }

        /// <summary>Return a sorted copy of <paramref name="arr"/>.</summary>
        public static int[] Sorted(int[] arr)
        {
            var copy = (int[])arr.Clone();
            Sort(copy);
            return copy;
        }

        // ─────────────────────────────────────────────────────────────────────
        // Internal implementation
        // ─────────────────────────────────────────────────────────────────────

        private static void SortImpl(int[] a, int lo, int hi, int depth)
        {
            while (lo < hi)
            {
                int size = hi - lo + 1;

                if (depth <= 0 || size <= SmallN)
                {
                    InsertionSort(a, lo, hi);
                    return;
                }

                // Counting sort for dense integer subarrays
                int mn = a[lo], mx = a[lo];
                for (int i = lo + 1; i <= hi; i++)
                {
                    if (a[i] < mn) mn = a[i];
                    if (a[i] > mx) mx = a[i];
                }
                long span = (long)mx - mn;
                if (span < (long)size * 4)
                {
                    int[] counts = new int[span + 1];
                    for (int i = lo; i <= hi; i++) counts[a[i] - mn]++;
                    int k = lo;
                    for (long v = 0; v <= span; v++)
                        for (int c = 0; c < counts[v]; c++)
                            a[k++] = (int)v + mn;
                    return;
                }

                // Monotone checks
                if (a[lo] <= a[lo + 1] && a[lo + 1] <= a[lo + 2])
                {
                    bool asc = true;
                    for (int i = lo; i < hi; i++) if (a[i] > a[i + 1]) { asc = false; break; }
                    if (asc) return;

                    bool desc = true;
                    for (int i = lo; i < hi; i++) if (a[i] < a[i + 1]) { desc = false; break; }
                    if (desc) { Reverse(a, lo, hi); return; }
                }

                // Oracle-seeded golden-ratio pivot positions
                double absC = Rng.NextDouble();
                if (absC == 0.0) absC = 1e-15;
                int sp   = hi - lo;
                int idx1 = lo + (int)(sp * Phi2 * absC);
                int idx2 = lo + (int)(sp * Phi  * absC);

                int p1 = Ninther(a, lo, hi, idx1);
                int p2 = Ninther(a, lo, hi, idx2);

                DualPartition(a, lo, hi, p1, p2, out int lt, out int gt);

                int leftN  = lt - lo;
                int midN   = gt - lt + 1;
                int rightN = hi - gt;

                // Sort smallest two regions; tail-call on largest
                SortRegions(a, lo, lt - 1, lt, gt, gt + 1, hi,
                            leftN, midN, rightN, depth - 1,
                            out int nextLo, out int nextHi);

                lo = nextLo; hi = nextHi; depth--;
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int Ninther(int[] a, int lo, int hi, int idx)
        {
            int i0 = Math.Max(lo, idx - 1);
            int i2 = Math.Min(hi, idx + 1);
            int x = a[i0], y = a[idx], z = a[i2];
            if (x > y) { int t = x; x = y; y = t; }
            if (y > z) { int t = y; y = z; z = t; }
            if (x > y) { int t = x; x = y; y = t; }
            return y;
        }

        private static void DualPartition(
            int[] a, int lo, int hi, int p1, int p2,
            out int lt, out int gt)
        {
            if (p1 > p2) { int t = p1; p1 = p2; p2 = t; }
            lt = lo; gt = hi;
            int i = lo;
            while (i <= gt)
            {
                int v = a[i];
                if (v < p1)
                {
                    int t = a[lt]; a[lt] = a[i]; a[i] = t;
                    lt++; i++;
                }
                else if (v > p2)
                {
                    int t = a[i]; a[i] = a[gt]; a[gt] = t;
                    gt--;
                }
                else i++;
            }
        }

        private static void SortRegions(
            int[] a,
            int lo1, int hi1, int lo2, int hi2, int lo3, int hi3,
            int n1, int n2, int n3, int depth,
            out int tailLo, out int tailHi)
        {
            // Identify smallest two; recurse into them; tail-call on largest
            if (n1 <= n2 && n1 <= n3)
            {
                if (n2 <= n3) { RecurseTwo(a, lo1, hi1, lo2, hi2, depth); tailLo = lo3; tailHi = hi3; }
                else          { RecurseTwo(a, lo1, hi1, lo3, hi3, depth); tailLo = lo2; tailHi = hi2; }
            }
            else if (n2 <= n1 && n2 <= n3)
            {
                if (n1 <= n3) { RecurseTwo(a, lo2, hi2, lo1, hi1, depth); tailLo = lo3; tailHi = hi3; }
                else          { RecurseTwo(a, lo2, hi2, lo3, hi3, depth); tailLo = lo1; tailHi = hi1; }
            }
            else
            {
                if (n1 <= n2) { RecurseTwo(a, lo3, hi3, lo1, hi1, depth); tailLo = lo2; tailHi = hi2; }
                else          { RecurseTwo(a, lo3, hi3, lo2, hi2, depth); tailLo = lo1; tailHi = hi1; }
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static void RecurseTwo(int[] a, int lo1, int hi1, int lo2, int hi2, int depth)
        {
            if (lo1 < hi1) SortImpl(a, lo1, hi1, depth);
            if (lo2 < hi2) SortImpl(a, lo2, hi2, depth);
        }

        private static void InsertionSort(int[] a, int lo, int hi)
        {
            for (int i = lo + 1; i <= hi; i++)
            {
                int key = a[i], j = i - 1;
                while (j >= lo && a[j] > key) { a[j + 1] = a[j]; j--; }
                a[j + 1] = key;
            }
        }

        private static void Reverse(int[] a, int lo, int hi)
        {
            while (lo < hi) { int t = a[lo]; a[lo] = a[hi]; a[hi] = t; lo++; hi--; }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int BitLength(int n)
        {
            int bits = 0;
            while (n > 0) { n >>= 1; bits++; }
            return bits - 1;
        }
    }
}

File: src/typescript/logos_sort.ts
/**
 * LogosSort — Golden-ratio dual-pivot introsort (TypeScript)
 *
 * Two pivots placed at φ (≈61.8%) and 1−φ (≈38.2%) golden-ratio positions
 * of a chaos-seeded index. Ninther pivot refinement, three-way partition,
 * counting-sort fast path for dense integer ranges, O(log n) depth limit.
 */

const PHI  = 0.6180339887498949;
const PHI2 = 0.3819660112501051;
const SMALL_N = 48;

function ninther(a: number[], lo: number, hi: number, idx: number): number {
  const i0 = Math.max(lo, idx - 1);
  const i2 = Math.min(hi, idx + 1);
  let x = a[i0], y = a[idx], z = a[i2];
  if (x > y) { [x, y] = [y, x]; }
  if (y > z) { [y, z] = [z, y]; }
  if (x > y) { [x, y] = [y, x]; }
  return y;
}

function dualPartition(
  a: number[], lo: number, hi: number, p1: number, p2: number
): [number, number] {
  if (p1 > p2) { [p1, p2] = [p2, p1]; }
  let lt = lo, gt = hi, i = lo;
  while (i <= gt) {
    const v = a[i];
    if (v < p1) {
      [a[lt], a[i]] = [a[i], a[lt]];
      lt++; i++;
    } else if (v > p2) {
      [a[i], a[gt]] = [a[gt], a[i]];
      gt--;
    } else {
      i++;
    }
  }
  return [lt, gt];
}

function sortImpl(a: number[], lo: number, hi: number, depth: number): void {
  while (lo < hi) {
    const size = hi - lo + 1;

    if (depth <= 0 || size <= SMALL_N) {
      const sub = a.slice(lo, hi + 1).sort((x, y) => x - y);
      for (let i = 0; i < sub.length; i++) a[lo + i] = sub[i];
      return;
    }

    // Counting sort for dense integer subarrays
    if (Number.isInteger(a[lo])) {
      let mn = a[lo], mx = a[lo];
      for (let i = lo + 1; i <= hi; i++) {
        if (a[i] < mn) mn = a[i];
        if (a[i] > mx) mx = a[i];
      }
      const span = mx - mn;
      if (span < size * 4) {
        const counts = new Int32Array(span + 1);
        for (let i = lo; i <= hi; i++) counts[a[i] - mn]++;
        let k = lo;
        for (let v = 0; v <= span; v++) {
          for (let c = 0; c < counts[v]; c++) a[k++] = v + mn;
        }
        return;
      }
    }

    // Monotone checks
    if (a[lo] <= a[lo + 1] && a[lo + 1] <= a[lo + 2]) {
      let asc = true;
      for (let i = lo; i < hi; i++) if (a[i] > a[i + 1]) { asc = false; break; }
      if (asc) return;

      let desc = true;
      for (let i = lo; i < hi; i++) if (a[i] < a[i + 1]) { desc = false; break; }
      if (desc) {
        for (let l = lo, r = hi; l < r; l++, r--) [a[l], a[r]] = [a[r], a[l]];
        return;
      }
    }

    // Oracle-seeded golden-ratio pivot positions
    const absC = Math.abs(Math.random() * 2 - 1) || 1e-15;
    const sp   = hi - lo;
    const idx1 = lo + Math.floor(sp * PHI2 * absC);
    const idx2 = lo + Math.floor(sp * PHI  * absC);

    const p1 = ninther(a, lo, hi, idx1);
    const p2 = ninther(a, lo, hi, idx2);

    const [lt, gt] = dualPartition(a, lo, hi, p1, p2);

    const regions: [number, number, number][] = [
      [lt - lo,      lo,    lt - 1],
      [gt - lt + 1,  lt,    gt    ],
      [hi - gt,      gt + 1, hi   ],
    ].sort((a, b) => a[0] - b[0]) as [number, number, number][];

    for (let i = 0; i < 2; i++) {
      const [, rLo, rHi] = regions[i];
      if (rLo < rHi) sortImpl(a, rLo, rHi, depth - 1);
    }

    [, lo, hi] = regions[2];
    depth--;
  }
}

/**
 * Sort an array of numbers in-place using LogosSort.
 * Returns the same array sorted.
 */
export function logosSort(arr: number[]): number[] {
  const n = arr.length;
  if (n < 2) return arr;
  const depthLimit = 2 * Math.floor(Math.log2(n)) + 4;
  sortImpl(arr, 0, n - 1, depthLimit);
  return arr;
}
